Add Code Insights team page (#2140)
sourcegraph/about: handbook/engineering/code-insights/index.md

> # Code insights team (coming soon)

Reviewer: My personal preference is to not prematurely create a team page like this until the team exists. The team can't exist until we have a PM. Can we inline this into the web team?

Author: This information (at least the goals and mission statement) is currently inlined in the web team page, but @jeanduplessis and I were worried that it is burried too deep in there for a candidate (who came to the job listing) to find it and can't do a good job to get them excited specifically for this team (because they for example have an interest in this mission specifically, or bring unique experience and skills for it).

The next step would be to link to this page directly from the "These teams are hiring for this role" section in the job listings, instead of just listing "Web". I am worried that "Web" isn't enticing to someone who doesn't want to join a (perceived) general-purpose frontend team - but 
 our split into dedicated product (and platform) teams is right around the corner.
And in the context of the backend engineer job listing, "Web" is very confusing to list (it's not enticing from the name, and a candidate would have to read between the lines of the web team page to imagine what they would even be working on).

Reviewer 2: @nicksnyder I had direct feedback from a candidate that they didn't really understand what the Code Insights team was about when they tried preparing for their interview. I'm not against inlining it in the Web team page, but as Felix mentioned, that page can quickly start getting busy and we need to do a better job of selling the Code Insights team for possible candidates.

Also considering we are hiring for code insights (both backend and frontend) and have a PM and EM lined up (Joel & Felix) and internally in the Web team we are already starting to treat it as a separate team, I don't think it's that premature to have a team page up.

Reviewer 3: Agree with @jeanduplessis - this is a team that is set up well and I think it makes sense to start to separate some of the team info.

Reviewer: I am fine with this if you think this is the best candidate experience for hiring. 

I suspect the candidate's concerns could also be addressed by simply organizing the information on the web team page better (which should probably happen regardless).

I want to ensure we don't prematurely fragment a single team in our documentation or in practice. Can you please chat with @lguychard to understand how he is approaching the same problem on the search team.

Reviewer 2: @nicksnyder Felix and I chatted with Loïc and we are confident that we should move forward with a separate page.

Forming the Code Insights team is not something that is far off in our roadmap, in fact, Joel and Felix are actively working on this as their primary focus. To make this more explicit I will update the Web team page to indicate that Joel is the acting PM for the team until we hire a new dedicated PM.

Code Insights also differs from Web Platform and Extensions & Integrations in that it is much more of a classical product engineering team and as such, I think it is good to separate it out from the rest of the Web team in terms of how we position it to candidates.

We will also move ahead and set up a dedicated frontend engineer job spec for this team and have a separate hiring pipeline from the Web team for it, lead by Felix.

Reviewer: So to be clear, are we ready to declare the code insights team as a separate team that has an EM (Felix) and a PM (Joel) where we are going to be actively hiring for? If all those are true, then let's lean in to this and be more explicit. Make this look like a complete team page, just like any other team.

Reviewer: And also
- Update the growth plan on the web team page to pull out code insights since this PR is doing exactly that split already.
- Remove code insights goals from web team

If any of the above feels uncomfortable, then we aren't ready to document a new team yet and we should instead focus in improving the existing web team's docs to solve the candidate problem.

Reviewer: Please review this: https://github.com/sourcegraph/about/pull/2153

Reviewer 4: I'll let others comment, but I'm comfortable splitting code insights out into its own handbook pages at this point, and think that may help clear up some confusion with the page organization. I'll leave some other comments on #2153 about the web team's overall focus.

Reviewer 2: I agree we should lean in and commit to the formation of this team and create a distinction from the web team. 

@felixfbecker @Joelkw let's update this PR so the new page has all the relevant info needed.

> # Code insights team (coming soon)

Reviewer 2: ```suggestion
# Code insights team
```

I suggest we take the coming soon away, it doesn't add any value.

Reviewer: I agree

<img src="./screenshot.svg" alt="Screenshot of a code insights dashboard with graphs" />

## What is code insights?

Code insights is the first feature in Sourcegraph that can tell you things about your code at a **high level**. Here's a [demo](https://www.youtube.com/watch?v=XqeRb6Mc4Co) of a code insights prototype.

Sourcegraph is in the unique position to give these insights because we have universal code search: To know _anything_ about your code at a high level means you must know _everything_ about your code at a low level.

Code insights connects a lot of features that Sourcegraph already has and builds on top of them.
We go beyond single-step code intelligence and search to connect the full cycle of analyzing (code intelligence), monitoring (code insights), and actionably changing a codebase (campaigns).

Code insights will also allow to combine these Sourcegraph-created metrics with other third-party data sources.

Code insights is the first feature primarily built for non-search-based user personas.

> ## Goals

Reviewer: I am hesitant to label these as goals because @christinaforney and I are working hard to align the team on what "goals" mean and this does not fit that definition.

Here are examples of what we want "goals" to be:
- https://about.sourcegraph.com/handbook/engineering/cloud/goals
- https://about.sourcegraph.com/handbook/engineering/code-intelligence/goals

Author: This section was taken directly from the web team's Goal page: https://about.sourcegraph.com/handbook/engineering/web/goals#medium-term

Reviewer 3: I've already spoken with @Joelkw and the team about goals, he was more ahead and pushed a nice update to goals right before we started to refine our format. He will be coming back to this to align on format with the team.

Separately, I would only move goals here that are specific to code insights and explicitly remove them from the web team goals to avoid duplication. Perhaps an interim solution is to just link back to the web team goals?

Reviewer: ```suggestion
## Vision
```

Can we re-brand this as the "vision" for the team?

1. Expose the value of Sourcegraph's knowledge of your codebase to users at all levels of an organization.
1. Expose generalizable metrics that let our users measure and track their own goals, whether those are migrations, code smells, test coverage,  security needs, cross-collaboration, or other information about code.
1. Make these metrics scalable across thousands of repositories on a Sourcegraph instance.
1. Expose features that let our users build their own insights.
1. Expose features that let users drill down into insights by teams, repositories, or other sub-metrics.
1. Connect code insights to other Sourcegraph features – make a campaign to take action on a metric or set up code monitoring to notify you about changes to an insight.
1. Add plug-and-play integrations to Sourcegraph code insights that let users easily integrate third-party data into their insights.
1. Discover new, validated code metrics that engineering teams should track (and let teams track them).

## We're hiring for this team!

> We are building the code insights team by hiring [frontend engineers](../hiring/software-engineer-frontend.md) and [backend engineers](../hiring/software-engineer-backend.md).

Reviewer: @christinaforney @Joelkw What is the plan for the PM of this team. Is it you Joel, or are we going to hire for this?

Reviewer 4: @nicksnyder the plan is that I will be the code insights PM. (In any interim, I will also continue to be the web team PM until we hire a PM for the web team.)

Reviewer: This is valuable information to put in the handbook

> We are building the code insights team by hiring [frontend engineers](../hiring/software-engineer-frontend.md) and [backend engineers](../hiring/software-engineer-backend.md).

Reviewer 4: Sounds good. 

```suggestion
We are building the code insights team by hiring [frontend engineers](../hiring/software-engineer-frontend.md) and [backend engineers](../hiring/software-engineer-backend.md). The engineering manager for this team will be [Felix Becker](../../../company/team/index.md#felix-becker) and the product manager will be [Joel Kwartler](../../../company/team/index.md#joel-kwartler-he-him).
```

# Code insights team

> The code insights team will be responsible for building and delivering [**code insights**](https://docs.google.com/document/d/1EHzor6I1GhVVIpl70mH-c10b1tNEl_p1xRMJ9qHQfoc/edit) to engineering leaders, empowering data-driven decisions in engineering organizations.

Reviewer: I think it would be better to inline more of a definition of what code insights is, rather than linking out to an RFC (here) and a video (in the next section). Those are great things to go read to learn more, but I would really like to see a crisp ~3-5 sentence self-contained description of code insights right here.

<img src="./screenshot.svg" alt="Screenshot of a code insights dashboard with graphs" />

## What is code insights?

Code insights is the first feature in Sourcegraph that can tell you things about your code at a **high level**. Here's a [demo](https://www.youtube.com/watch?v=XqeRb6Mc4Co) of a code insights prototype.

Sourcegraph is in the unique position to give these insights because we have universal code search: To know _anything_ about your code at a high level means you must know _everything_ about your code at a low level.

Code insights connects a lot of features that Sourcegraph already has and builds on top of them.
We go beyond single-step code intelligence and search to connect the full cycle of analyzing (code intelligence), monitoring (code insights), and actionably changing a codebase (campaigns).

Code insights will also allow to combine these Sourcegraph-created metrics with other third-party data sources.

Code insights is the first feature primarily built for non-search-based user personas.

## Goals 

> **Problem:** Code insights solves two problems. 
> 
> 1. The Sourcegraph sales cycle is unusual because although we consistently wow our users, the economic decision-maker is usually not one of these users. Instead, the people with the power to sign a contract with us are higher up within an organization and usually depend on running a trial to fully understand the value of Sourcegraph. If code insights answers these personas' higher-level questions about codebases that our core features do not currently answer, it could dramatically speed up our sales cycle as well as our sales pipeline.
> 
> 1. Sourcegraph has lots of information about your code, but currently only exposes specific, low-level information at the file or directory level. Engineers and engineering managers/directors/VPs have lots of questions about their codebase that Sourcegraph can answer but currently does not. 

Reviewer: my opinion...

Problem statement should be framed from a user perspective of "I want X, but can't get X today (from Sourcegraph or any other tool)."

Value to Sourcegraph is worth calling out, but should be distinctly separate.

```suggestion
**Problem:** Engineers and engineering managers/directors/VPs want to be able to understand their codebase at a high level (which parts of the code base are health/unhealthy? where are engineers spending their time editing/reading code?). Existing tools like X, Y, Z, don't answer these questions because A, B, Z.

**Value to Sourcegraph:** The Sourcegraph sales cycle is unusual because although we consistently wow our users, the economic decision-maker is usually not one of these users. Instead, the people with the power to sign a contract with us are higher up within an organization and usually depend on running a trial to fully understand the value of Sourcegraph. If code insights answers these personas' higher-level questions about codebases that our core features do not currently answer, it could dramatically speed up our sales cycle as well as our sales pipeline by giving decision makers a reason to buy Sourcegraph without needing to run a trial.
```

**Milestones:**

_Because code insights is an entirely new feature and closely informed by customer feedback, the further in the future goals get, the more flexible the order of these milestones is (especially when it comes to interleaving "business" milestones like beta/GA/paid with feature milestones)._ 

> 1. 🔄 Code insights prototypes let us gather qualitative feedback to guide the initial product direction. 

Reviewer: when is this "done"?

My proposal would be that we have 3 customers who are actively using code insights. And when I say active here it is a qualitative thing (they are giving us feedback and saying that they are using it). Next milestone is to make this quantitative.

Reviewer 4: That works; agree it could be clearer.

> 1. We can quantitatively measure the adoption of code insights prototypes. 

Reviewer: ```suggestion
1. We have decided on and implemented a metric to quantitatively measure the adoption of code insights prototypes. 
```

Reviewer 4: (Made this change and linked the RFC as well.)

1. ✅ Three customers give us qualitative feedback after using our code insights prototypes to guide the initial product direction. 

    - **Outcome:** We have a list of potential features and their likely value that we can use to achieve our initial adoption milestones ([Productboard view](https://sourcegraph.productboard.com/feature-board/2327586-code-insights-next-objective)). 

1. 🔄 We have decided on and implemented metrics to quantitatively measure the adoption of code insights prototypes (see [RFC 279](https://docs.google.com/document/d/1I10tm5CFZvzQYNeV--JacRGLLIUesXQBp6ZO8uhakRs/edit#)). 

    - **Outcome:** We have weekly quantitative reports on the use of code insights at each customer. 

1. Code insights can scale to large (if not the largest) codebases. 

    - **Outcome:** All but the largest customers can use insights prototypes over their entire codebase to answer questions about the state of _all_ of their code.

> 1. Code insights provides the functionality and positioning to be immediately useful to many customers (this may include integrations with campaigns, code monitoring, or third party data sources – see [Productboard feature view](https://sourcegraph.productboard.com/feature-board/1793095-code-insights)).  

Reviewer: ```suggestion
1. Code insights are useful to a broad set of customers (this may include integrations with campaigns, code monitoring, or third party data sources – see [Productboard feature view](https://sourcegraph.productboard.com/feature-board/1793095-code-insights)).  
```

    - **Outcome:** Code insights features are actively in use by VPs/Directors at 5+ enterprise customers.  

> 1. Customers can easily create their own custom insights in an intentionally-designed UI.

Reviewer: > in an intentionally-designed UI.

What does this mean?

Reviewer 4: I should have said "dedicated UI" for clarity, even if the this milestone's UI form won't be as "dedicated" as something like the campaigns UI yet.

Reviewer: So like a WYSIWYG editor for creating code insights?

Reviewer 4: Sort of, though I don't know what WYSIWYG means when WYS is code and WYG is a visual graph (is Looker/Bigquery what you mean?) because it might still be code-as-UI here. The core of this is "you don't have to go and set it up like a prototype in your settings file" but "there's a UI that we built for this purpose" even if this isn't its final form as we learn about problems or personas who would prefer a different UI. 

The simplest first step would be to just pull out the code insights settings code into its own settings-like editor you access somewhere in the UI so you don't get distracted or confused by a busy interface when setting up an insight.

>     - **Outcome:** Code insights can move from prototype to beta feature (easy to enable or enabled by default, documented, and supported by the CE team rather than the product team).  

Reviewer: Should we have an outcome like "We have at least one customer created code insight that is actively used"?

Reviewer 4: Yep, great thought.

> 1. Code insights is GA (generally available) for all customers and has low/no barriers to wide adoption. 

Reviewer: Should we have an outcome of "x% of sourcegraph customers have active usage of code insights"? (this would prove broad appeal)

Reviewer 4: Yes, I think that's a better, more specific way of capturing the second outcome (L55) so I'll modify that one.

Code insights is the first feature primarily built for non-search-based user personas.

## Goals 

**Problem:** Code insights solves two problems. 

> 1. The Sourcegraph sales cycle is unusual because although we consistently wow our users, the economic decision-maker is usually not one of these users. Instead, the people with the power to sign a contract with us are higher up within an organization and usually depend on running a trial to fully understand the value of Sourcegraph. If code insights answers these personas' higher-level questions about codebases that our core features do not currently answer, it could dramatically speed up our sales cycle as well as our sales pipeline. These personas . 

Reviewer 2: There's. a trailing "These personas . " - not sure if that shouldn't be there or is an incomplete sentence.

Reviewer 4: Ah, leftover from an edit – cleaned up, thank you!
